File: numberOfIslands/numberOfIslands.cpp
#include <vector>
#include <set>
#include <utility>
class Solution {
    //DFS, when you encounter an island, check every piece of land connected to it, and
    //mark it as visited so you dont recount it. Need to check every direction since
    //in the case of an island
    /**
     * 000110
     * 001111
     * 000111
     * 
     * The first 1 in row 2 would count as its own island
    */
public:
    int numIslands(std::vector<std::vector<char>>& grid) {
        std::set<std::pair<int, int>> visited;
        int count = 0;
        for(int i = 0; i < grid.size(); ++i)
        {
            for(int j = 0; j < grid[0].size(); ++j)
            {
                std::pair<int, int> coords = std::make_pair(i, j);
                if(visited.count(coords) == 0 && grid[i][j] == '1')
                {
                    ++count;
                    islandHelper(grid, i, j, visited);
                }
                else
                {
                    visited.insert(coords);
                }
            }
        }
        return count;
    }
    
    void islandHelper(std::vector<std::vector<char>>& grid, int i, int j, std::set<std::pair<int,int>>& visited)
    {
        if(i < 0 || i >= grid.size())
            return;
        if(j < 0 || j >= grid[0].size())
            return;
        std::pair<int,int> coords = std::make_pair(i, j);
        if(visited.count(coords) == 0 && grid[i][j] == '1')
        {
            visited.insert(coords);
            islandHelper(grid, i, j+1, visited);
            islandHelper(grid, i+1, j, visited);
            islandHelper(grid, i, j-1, visited);
            islandHelper(grid, i-1, j, visited);
        }
    }
};

File: numberOfIslands/numberOfIslands.py
from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        islands = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if(grid[i][j] == '1'):
                    self._dfs(i, j, grid)
                    islands += 1
        return islands
    
    def _dfs(self, i: int, j: int, grid: List[List[str]]) -> None:
        if(i < 0 or i >= len(grid) or 
           j < 0 or j >= len(grid[0]) or
           grid[i][j] == '0'):
            return
        grid[i][j] = '0'
        self._dfs(i+1, j, grid)
        self._dfs(i-1, j, grid)        
        self._dfs(i, j+1, grid)        
        self._dfs(i, j-1, grid)      